cpp/src/groupby/sort/group_single_pass_reduction_util.cuh

```diff
@@ -21,20 +21,133 @@
 #include <cudf/column/column_view.hpp>
 #include <cudf/detail/aggregation/aggregation.cuh>
 #include <cudf/detail/iterator.cuh>
-#include <cudf/detail/utilities/device_atomics.cuh>
 #include <cudf/detail/valid_if.cuh>
-#include <cudf/table/table_device_view.cuh>
 #include <cudf/types.hpp>
 #include <cudf/utilities/span.hpp>
 
 #include <rmm/cuda_stream_view.hpp>
 #include <rmm/exec_policy.hpp>
 
+#include <thrust/functional.h>
+#include <thrust/iterator/counting_iterator.h>
 #include <thrust/iterator/discard_iterator.h>
+#include <thrust/iterator/transform_output_iterator.h>
+#include <thrust/iterator/zip_iterator.h>
+#include <thrust/reduce.h>
 
 namespace cudf {
 namespace groupby {
 namespace detail {
+
+// ArgMin binary operator with tuple of (value, index)
+template <typename T>
+struct ArgMin {
+  CUDA_HOST_DEVICE_CALLABLE auto operator()(thrust::tuple<T, size_type> const& lhs,
+                                            thrust::tuple<T, size_type> const& rhs) const
+  {
+    if (thrust::get<1>(lhs) == cudf::detail::ARGMIN_SENTINEL)
+      return rhs;
+    else if (thrust::get<1>(rhs) == cudf::detail::ARGMIN_SENTINEL)
+      return lhs;
+    else
+      return thrust::get<0>(lhs) < thrust::get<0>(rhs) ? lhs : rhs;
+  }
+};
+
+// ArgMax binary operator with tuple of (value, index)
+template <typename T>
+struct ArgMax {
+  CUDA_HOST_DEVICE_CALLABLE auto operator()(thrust::tuple<T, size_type> const& lhs,
+                                            thrust::tuple<T, size_type> const& rhs) const
+  {
+    if (thrust::get<1>(lhs) == cudf::detail::ARGMIN_SENTINEL)
+      return rhs;
+    else if (thrust::get<1>(rhs) == cudf::detail::ARGMIN_SENTINEL)
+      return lhs;
+    else
+      return thrust::get<0>(lhs) > thrust::get<0>(rhs) ? lhs : rhs;
+  }
+};
+
+struct get_tuple_second_element {
+  template <typename T>
+  __device__ size_type operator()(thrust::tuple<T, size_type> const& rhs) const
+  {
+    return thrust::get<1>(rhs);
+  }
+};
+
+/**
+ * @brief Functor to store the boolean value to null mask.
+ */
+struct bool_to_nullmask {
+  mutable_column_device_view d_result;
+  __device__ void operator()(size_type i, bool rhs)
+  {
+    if (rhs) {
+      d_result.set_valid(i);
+    } else {
+      d_result.set_null(i);
+    }
+  }
+};
+
+/**
+ * @brief Returns index for non-null element, and SENTINEL for null element in a column.
+ *
+ */
+struct null_as_sentinel {
+  column_device_view const col;
+  size_type const SENTINEL;
+  __device__ size_type operator()(size_type i) const { return col.is_null(i) ? SENTINEL : i; }
+};
```

Comment:
Can't `null_replacement_iterator` be used instead? https://github.com/rapidsai/cudf/blob/dfe0a03d40f956561e9416b6cc00246d99537aa5/cpp/include/cudf/detail/iterator.cuh#L162

Reply:
`null_replacement_iterator` returns values of the column. Here, indices are needed.